### PR TITLE
init project

### PR DESCRIPTION
**Pull request by DaveXiong**

### CODE REVIEW

File: workflow-java-sdk-process-plugin/src/main/java/com/consoleconnect/sdk/workflow/process/plugin/ProcessParseListener.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import org.camunda.bpm.engine.delegate.TaskListener;
+import org.camunda.bpm.engine.impl.bpmn.behavior.UserTaskActivityBehavior;
+import org.camunda.bpm.engine.impl.bpmn.parser.AbstractBpmnParseListener;
+import org.camunda.bpm.engine.impl.bpmn.parser.BpmnParseListener;
+import org.camunda.bpm.engine.impl.pvm.delegate.ActivityBehavior;
+import org.camunda.bpm.engine.impl.pvm.process.ActivityImpl;
+import org.camunda.bpm.engine.impl.pvm.process.ScopeImpl;
+import org.camunda.bpm.engine.impl.util.xml.Element;
+
+public class ProcessParseListener extends AbstractBpmnParseListener implements BpmnParseListener {
```
Comment (dxiongtech) [inline workflow-java-sdk-process-plugin/src/main/java/com/consoleconnect/sdk/workflow/process/plugin/ProcessParseListener.java]: MINOR Document this public class by adding an explicit description. [rule](https://sq.ci.consoleconnect.com/coding_rules#rule_key=squid%3AUndocumentedApi)

File: workflow-java-sdk-process-plugin/src/main/java/com/consoleconnect/sdk/workflow/process/plugin/ProcessIncidentHandler.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+   * @see org.camunda.bpm.engine.impl.incident.IncidentHandler#handleIncident(org.
+   * camunda.bpm.engine.impl.incident.IncidentContext, java.lang.String)
+   */
+  @Override
+  public void handleIncident(IncidentContext context, String message) {
+    super.handleIncident(context, message);
+
+    LOGGER.log(Level.INFO, () -> "handleIncident:" + context + message);
+
+    // TODO: FIXME: send email
```
Comment (dxiongtech) [inline workflow-java-sdk-process-plugin/src/main/java/com/consoleconnect/sdk/workflow/process/plugin/ProcessIncidentHandler.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://sq.ci.consoleconnect.com/coding_rules#rule_key=squid%3AS1134)
INFO Complete the task associated to this TODO comment. [rule](https://sq.ci.consoleconnect.com/coding_rules#rule_key=squid%3AS1135)

File: workflow-java-sdk-process-plugin/src/main/java/com/consoleconnect/sdk/workflow/process/plugin/ProcessIncidentHandler.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+   * @see org.camunda.bpm.engine.impl.incident.IncidentHandler#resolveIncident(org.
+   * camunda.bpm.engine.impl.incident.IncidentContext)
+   */
+  @Override
+  public void resolveIncident(IncidentContext context) {
+    super.resolveIncident(context);
+
+    LOGGER.log(Level.INFO, () -> "resolveIncident:" + context);
+
+    // TODO: FIXME: send email
```
Comment (dxiongtech) [inline workflow-java-sdk-process-plugin/src/main/java/com/consoleconnect/sdk/workflow/process/plugin/ProcessIncidentHandler.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://sq.ci.consoleconnect.com/coding_rules#rule_key=squid%3AS1134)
INFO Complete the task associated to this TODO comment. [rule](https://sq.ci.consoleconnect.com/coding_rules#rule_key=squid%3AS1135)

File: workflow-java-sdk-process-plugin/src/main/java/com/consoleconnect/sdk/workflow/process/plugin/ProcessIncidentHandler.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+   * 
+   * @see org.camunda.bpm.engine.impl.incident.IncidentHandler#deleteIncident(org.
+   * camunda.bpm.engine.impl.incident.IncidentContext)
+   */
+  @Override
+  public void deleteIncident(IncidentContext context) {
+    super.deleteIncident(context);
+
+    LOGGER.log(Level.INFO, () -> "deleteIncident:" + context);
+    // TODO: FIXME: send email
```
Comment (dxiongtech) [inline workflow-java-sdk-process-plugin/src/main/java/com/consoleconnect/sdk/workflow/process/plugin/ProcessIncidentHandler.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://sq.ci.consoleconnect.com/coding_rules#rule_key=squid%3AS1134)
INFO Complete the task associated to this TODO comment. [rule](https://sq.ci.consoleconnect.com/coding_rules#rule_key=squid%3AS1135)

File: workflow-java-sdk-process-plugin/src/main/java/com/consoleconnect/sdk/workflow/process/plugin/ProcessIncidentHandler.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+/**
+ * @author dxiong
+ *
+ */
+public class ProcessIncidentHandler extends DefaultIncidentHandler {
+
+  private static final Logger LOGGER = Logger.getLogger(ProcessIncidentHandler.class.getName());
+
+  public ProcessIncidentHandler(String type) {
```
Comment (dxiongtech) [inline workflow-java-sdk-process-plugin/src/main/java/com/consoleconnect/sdk/workflow/process/plugin/ProcessIncidentHandler.java]: MINOR Document this public constructor by adding an explicit description. [rule](https://sq.ci.consoleconnect.com/coding_rules#rule_key=squid%3AUndocumentedApi)

File: workflow-java-sdk-process-plugin/src/main/java/com/consoleconnect/sdk/workflow/process/plugin/ConsoleProcessEnginePlugin.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+import java.util.logging.Logger;
+
+import org.camunda.bpm.engine.ProcessEngine;
+import org.camunda.bpm.engine.impl.bpmn.parser.BpmnParseListener;
+import org.camunda.bpm.engine.impl.cfg.ProcessEngineConfigurationImpl;
+import org.camunda.bpm.engine.impl.cfg.ProcessEnginePlugin;
+import org.camunda.bpm.engine.impl.incident.IncidentHandler;
+import org.camunda.bpm.engine.runtime.Incident;
+
+public class ConsoleProcessEnginePlugin implements ProcessEnginePlugin {
```
Comment (dxiongtech) [inline workflow-java-sdk-process-plugin/src/main/java/com/consoleconnect/sdk/workflow/process/plugin/ConsoleProcessEnginePlugin.java]: MINOR Document this public class by adding an explicit description. [rule](https://sq.ci.consoleconnect.com/coding_rules#rule_key=squid%3AUndocumentedApi)

File: workflow-java-sdk-process-plugin/src/main/java/com/consoleconnect/sdk/workflow/process/plugin/ProcessTaskListener.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+  private static final Logger LOGGER = Logger.getLogger(ProcessTaskListener.class.getName());
+
+  @Override
+  public void notify(DelegateTask task) {
+    LOGGER.info("Event '" + task.getEventName() + "' received by Task Listener for Task:"
+        + " activityId=" + task.getTaskDefinitionKey() + ", name='" + task.getName() + "'"
+        + ", taskId=" + task.getId() + ", assignee='" + task.getAssignee() + "'"
+        + ", candidateGroups='" + task.getCandidates() + "'");
+
+    // TODO: send email
```
Comment (dxiongtech) [inline workflow-java-sdk-process-plugin/src/main/java/com/consoleconnect/sdk/workflow/process/plugin/ProcessTaskListener.java]: INFO Complete the task associated to this TODO comment. [rule](https://sq.ci.consoleconnect.com/coding_rules#rule_key=squid%3AS1135)

File: workflow-java-sdk-process-plugin/src/main/java/com/consoleconnect/sdk/workflow/process/plugin/ProcessTaskListener.java [hunk 8]
```diff
@@ -0,0 +1,22 @@
+package com.consoleconnect.sdk.workflow.process.plugin;
+
+import java.util.logging.Logger;
+
+import org.camunda.bpm.engine.delegate.DelegateTask;
+import org.camunda.bpm.engine.delegate.TaskListener;
+
+public class ProcessTaskListener implements TaskListener {
```
Comment (dxiongtech) [inline workflow-java-sdk-process-plugin/src/main/java/com/consoleconnect/sdk/workflow/process/plugin/ProcessTaskListener.java]: MINOR Document this public class by adding an explicit description. [rule](https://sq.ci.consoleconnect.com/coding_rules#rule_key=squid%3AUndocumentedApi)

**Comment by dxiongtech** [comment]
SonarQube analysis reported 11 issues
* ![MAJOR][MAJOR] 3 major
* ![MINOR][MINOR] 4 minor
* ![INFO][INFO] 4 info

Watch the comments in this conversation to review them.

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: workflow-java-sdk-process-plugin/src/main/java/com/consoleconnect/sdk/workflow/process/plugin/ProcessParseListener.java (hunk 1 above)
Comment (DaveXiong) [reply workflow-java-sdk-process-plugin/src/main/java/com/consoleconnect/sdk/workflow/process/plugin/ProcessParseListener.java]: ACK

File: workflow-java-sdk-process-plugin/src/main/java/com/consoleconnect/sdk/workflow/process/plugin/ProcessIncidentHandler.java (hunk 2 above)
Comment (DaveXiong) [reply workflow-java-sdk-process-plugin/src/main/java/com/consoleconnect/sdk/workflow/process/plugin/ProcessIncidentHandler.java]: ACK

File: workflow-java-sdk-process-plugin/src/main/java/com/consoleconnect/sdk/workflow/process/plugin/ProcessIncidentHandler.java (hunk 3 above)
Comment (DaveXiong) [reply workflow-java-sdk-process-plugin/src/main/java/com/consoleconnect/sdk/workflow/process/plugin/ProcessIncidentHandler.java]: ACK

File: workflow-java-sdk-process-plugin/src/main/java/com/consoleconnect/sdk/workflow/process/plugin/ProcessIncidentHandler.java (hunk 4 above)
Comment (DaveXiong) [reply workflow-java-sdk-process-plugin/src/main/java/com/consoleconnect/sdk/workflow/process/plugin/ProcessIncidentHandler.java]: ACK

File: workflow-java-sdk-process-plugin/src/main/java/com/consoleconnect/sdk/workflow/process/plugin/ProcessIncidentHandler.java (hunk 5 above)
Comment (DaveXiong) [reply workflow-java-sdk-process-plugin/src/main/java/com/consoleconnect/sdk/workflow/process/plugin/ProcessIncidentHandler.java]: ACK

File: workflow-java-sdk-process-plugin/src/main/java/com/consoleconnect/sdk/workflow/process/plugin/ConsoleProcessEnginePlugin.java (hunk 6 above)
Comment (DaveXiong) [reply workflow-java-sdk-process-plugin/src/main/java/com/consoleconnect/sdk/workflow/process/plugin/ConsoleProcessEnginePlugin.java]: ACK

File: workflow-java-sdk-process-plugin/src/main/java/com/consoleconnect/sdk/workflow/process/plugin/ProcessTaskListener.java (hunk 7 above)
Comment (DaveXiong) [reply workflow-java-sdk-process-plugin/src/main/java/com/consoleconnect/sdk/workflow/process/plugin/ProcessTaskListener.java]: ACK

File: workflow-java-sdk-process-plugin/src/main/java/com/consoleconnect/sdk/workflow/process/plugin/ProcessTaskListener.java (hunk 8 above)
Comment (DaveXiong) [reply workflow-java-sdk-process-plugin/src/main/java/com/consoleconnect/sdk/workflow/process/plugin/ProcessTaskListener.java]: ACK

**Comment by dxiongtech** [comment]
SonarQube analysis reported 1 issue
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: workflow-java-sdk-service-rest/src/main/java/com/consoleconnect/sdk/workflow/service/rest/Request.java [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+    this.setMethod(method);
+    return this;
+  }
+
+  public Request withEndPoint(String endPoint) {
+    this.setEndPoint(endPoint);
+    return this;
+  }
+
+  public Request withHeaders(Map<String, String> headers) {
```
Comment (dxiongtech) [inline workflow-java-sdk-service-rest/src/main/java/com/consoleconnect/sdk/workflow/service/rest/Request.java]: MINOR Document this public method by adding an explicit description. [rule](https://sq.ci.consoleconnect.com/coding_rules#rule_key=squid%3AUndocumentedApi)

File: workflow-java-sdk-service-rest/src/main/java/com/consoleconnect/sdk/workflow/service/rest/Request.java [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+    return new Request();
+  }
+
+
+  public Request withProvider(String provider) {
+    this.setServiceProvider(provider);
+    return this;
+  }
+
+  public Request withMethod(String method) {
```
Comment (dxiongtech) [inline workflow-java-sdk-service-rest/src/main/java/com/consoleconnect/sdk/workflow/service/rest/Request.java]: MINOR Document this public method by adding an explicit description. [rule](https://sq.ci.consoleconnect.com/coding_rules#rule_key=squid%3AUndocumentedApi)

File: workflow-java-sdk-service-rest/src/main/java/com/consoleconnect/sdk/workflow/service/rest/Request.java [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+    this.setEndPoint(endPoint);
+    return this;
+  }
+
+  public Request withHeaders(Map<String, String> headers) {
+    this.setHeaders(headers);
+    return this;
+  }
+
+  public Request withHeader(String key, String value) {
```
Comment (dxiongtech) [inline workflow-java-sdk-service-rest/src/main/java/com/consoleconnect/sdk/workflow/service/rest/Request.java]: MINOR Document this public method by adding an explicit description. [rule](https://sq.ci.consoleconnect.com/coding_rules#rule_key=squid%3AUndocumentedApi)

File: workflow-java-sdk-service-rest/src/main/java/com/consoleconnect/sdk/workflow/service/rest/Request.java [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+    this.setServiceProvider(provider);
+    return this;
+  }
+
+  public Request withMethod(String method) {
+    this.setMethod(HttpMethod.valueOf(method));
+    return this;
+  }
+  
+  public Request withMethod(HttpMethod method) {
```
Comment (dxiongtech) [inline workflow-java-sdk-service-rest/src/main/java/com/consoleconnect/sdk/workflow/service/rest/Request.java]: MINOR Document this public method by adding an explicit description. [rule](https://sq.ci.consoleconnect.com/coding_rules#rule_key=squid%3AUndocumentedApi)

File: workflow-java-sdk-service-rest/src/main/java/com/consoleconnect/sdk/workflow/service/rest/Request.java [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+  String endPoint = "";
+  Map<String, String> headers = Collections.synchronizedMap(new HashMap<String, String>());
+  String body = null;
+
+
+  private Request() {
+
+  }
+
+  public static Request build() {
```
Comment (dxiongtech) [inline workflow-java-sdk-service-rest/src/main/java/com/consoleconnect/sdk/workflow/service/rest/Request.java]: MINOR Document this public method by adding an explicit description. [rule](https://sq.ci.consoleconnect.com/coding_rules#rule_key=squid%3AUndocumentedApi)

File: workflow-java-sdk-service-rest/src/main/java/com/consoleconnect/sdk/workflow/service/rest/Request.java [hunk 14]
```diff
@@ -0,0 +1,109 @@
+package com.consoleconnect.sdk.workflow.service.rest;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.springframework.http.HttpMethod;
+
+public class Request {
```
Comment (dxiongtech) [inline workflow-java-sdk-service-rest/src/main/java/com/consoleconnect/sdk/workflow/service/rest/Request.java]: MINOR Document this public class by adding an explicit description. [rule](https://sq.ci.consoleconnect.com/coding_rules#rule_key=squid%3AUndocumentedApi)

File: workflow-java-sdk-service-rest/src/main/java/com/consoleconnect/sdk/workflow/service/rest/Request.java [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+    this.setHeaders(headers);
+    return this;
+  }
+
+  public Request withHeader(String key, String value) {
+    headers.put(key, value);
+    return this;
+  }
+
+  public Request withBody(String body) {
```
Comment (dxiongtech) [inline workflow-java-sdk-service-rest/src/main/java/com/consoleconnect/sdk/workflow/service/rest/Request.java]: MINOR Document this public method by adding an explicit description. [rule](https://sq.ci.consoleconnect.com/coding_rules#rule_key=squid%3AUndocumentedApi)

File: workflow-java-sdk-service-rest/src/main/java/com/consoleconnect/sdk/workflow/service/rest/Request.java [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+    this.setMethod(HttpMethod.valueOf(method));
+    return this;
+  }
+  
+  public Request withMethod(HttpMethod method) {
+    this.setMethod(method);
+    return this;
+  }
+
+  public Request withEndPoint(String endPoint) {
```
Comment (dxiongtech) [inline workflow-java-sdk-service-rest/src/main/java/com/consoleconnect/sdk/workflow/service/rest/Request.java]: MINOR Document this public method by adding an explicit description. [rule](https://sq.ci.consoleconnect.com/coding_rules#rule_key=squid%3AUndocumentedApi)

File: workflow-java-sdk-service-rest/src/main/java/com/consoleconnect/sdk/workflow/service/rest/Request.java [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+  private Request() {
+
+  }
+
+  public static Request build() {
+    return new Request();
+  }
+
+
+  public Request withProvider(String provider) {
```
Comment (dxiongtech) [inline workflow-java-sdk-service-rest/src/main/java/com/consoleconnect/sdk/workflow/service/rest/Request.java]: MINOR Document this public method by adding an explicit description. [rule](https://sq.ci.consoleconnect.com/coding_rules#rule_key=squid%3AUndocumentedApi)

File: workflow-java-sdk-service-rest/src/main/java/com/consoleconnect/sdk/workflow/service/rest/DummyRestServiceProvider.java [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+ * @author dxiong
+ *
+ */
+@Component
+public class DummyRestServiceProvider extends RestServiceProvider {
+
+  /**
+   * 
+   */
+  public DummyRestServiceProvider() {
```
Comment (dxiongtech) [inline workflow-java-sdk-service-rest/src/main/java/com/consoleconnect/sdk/workflow/service/rest/DummyRestServiceProvider.java]: MINOR Document this public constructor by adding an explicit description. [rule](https://sq.ci.consoleconnect.com/coding_rules#rule_key=squid%3AUndocumentedApi)

File: workflow-java-sdk-service-rest/src/main/java/com/consoleconnect/sdk/workflow/service/rest/RestServiceRuntimeException.java [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+   * 
+   */
+  private static final long serialVersionUID = -8147070477340275335L;
+
+  private final transient Request request;
+
+  /**
+   * 
+   */
+  public RestServiceRuntimeException(String message, Request request) {
```
Comment (dxiongtech) [inline workflow-java-sdk-service-rest/src/main/java/com/consoleconnect/sdk/workflow/service/rest/RestServiceRuntimeException.java]: MINOR Document this public constructor by adding an explicit description. [rule](https://sq.ci.consoleconnect.com/coding_rules#rule_key=squid%3AUndocumentedApi)

File: workflow-java-sdk-service-rest/src/main/java/com/consoleconnect/sdk/workflow/service/rest/RestConfiguration.java [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+/**
+ * @author dxiong
+ *
+ */
+@Configuration
+@Service
+@PropertySource("classpath:rest.properties")
+public class RestConfiguration {
+  @Bean
+  public RestTemplate restTemplate() {
```
Comment (dxiongtech) [inline workflow-java-sdk-service-rest/src/main/java/com/consoleconnect/sdk/workflow/service/rest/RestConfiguration.java]: MINOR Document this public method by adding an explicit description. [rule](https://sq.ci.consoleconnect.com/coding_rules#rule_key=squid%3AUndocumentedApi)

File: workflow-java-sdk-service-rest/src/main/java/com/consoleconnect/sdk/workflow/service/rest/RestClient.java [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+public class RestClient {
+
+  private static final Logger LOG = LoggerFactory.getLogger(RestClient.class);
+
+
+
+  @Autowired
+  private List<RestServiceProvider> serviceProviders;
+
+  public ResponseEntity<Object> execute(Request request) {
```
Comment (dxiongtech) [inline workflow-java-sdk-service-rest/src/main/java/com/consoleconnect/sdk/workflow/service/rest/RestClient.java]: MINOR Document this public method by adding an explicit description. [rule](https://sq.ci.consoleconnect.com/coding_rules#rule_key=squid%3AUndocumentedApi)

File: workflow-java-sdk-service-rest/src/main/java/com/consoleconnect/sdk/workflow/service/rest/RestServiceProvider.java [hunk 22]
```diff
@@ -0,0 +1,85 @@
...
+
+
+  public String getBaseUrl() {
+    return env.getProperty(String.format("rest.%s.baseUrl", provider));
+
+  }
+
+
+
+  public ResponseEntity<Object> execute(Request request) {
```
Comment (dxiongtech) [inline workflow-java-sdk-service-rest/src/main/java/com/consoleconnect/sdk/workflow/service/rest/RestServiceProvider.java]: MINOR Document this public method by adding an explicit description. [rule](https://sq.ci.consoleconnect.com/coding_rules#rule_key=squid%3AUndocumentedApi)

File: workflow-java-sdk-service-rest/src/main/java/com/consoleconnect/sdk/workflow/service/rest/RestServiceProvider.java [hunk 23]
```diff
@@ -0,0 +1,85 @@
...
+
+  private String provider;
+
+  @Autowired
+  private RestTemplate restTemplate;
+
+  @Autowired
+  private Environment env;
+  
+  public RestServiceProvider(String provider) {
```
Comment (dxiongtech) [inline workflow-java-sdk-service-rest/src/main/java/com/consoleconnect/sdk/workflow/service/rest/RestServiceProvider.java]: MINOR Document this public constructor by adding an explicit description. [rule](https://sq.ci.consoleconnect.com/coding_rules#rule_key=squid%3AUndocumentedApi)

File: workflow-java-sdk-service-rest/src/main/java/com/consoleconnect/sdk/workflow/service/rest/RestServiceProvider.java [hunk 24]
```diff
@@ -0,0 +1,85 @@
...
+    this.provider = provider;
+  }
+
+  @PostConstruct
+  public void initi() {
```
Comment (dxiongtech) [inline workflow-java-sdk-service-rest/src/main/java/com/consoleconnect/sdk/workflow/service/rest/RestServiceProvider.java]: MINOR Document this public method by adding an explicit description. [rule](https://sq.ci.consoleconnect.com/coding_rules#rule_key=squid%3AUndocumentedApi)

**Comment by dxiongtech** [comment]
SonarQube analysis reported 16 issues
* ![MINOR][MINOR] 16 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**